File: eml-parser/src/s3.ts
import aws from "aws-sdk";

/**
 * Message published by S3 when a new object is created. Sent into `new_email`
 * queue.
 *
 * https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
 */
export interface S3CreateEvent {
  Records: Array<{
    eventVersion: string;
    s3: {
      bucket: { name: string };
      object: { key: string; size: number };
    };
  }>;
}

export interface S3Object {
  id: string;
  body: string;
}

export async function fetchFromS3(event: S3CreateEvent): Promise<S3Object> {
  const { eventVersion, s3 } = event.Records.pop()!;

  console.log(
    `Received an S3 event version ${eventVersion} of size ${s3.object.size}.`
  );

  const bucketName = s3.bucket.name;
  const objectName = decodeURI(s3.object.key); // a.k.a. email id

  const params: aws.S3.Types.GetObjectRequest = {
    Bucket: bucketName,
    Key: objectName,
  };
  const { Body } = await new aws.S3().getObject(params).promise();

  return {
    id: objectName,
    body: (Body || "").toString(),
  };
}

export async function uploadHtmlToS3(
  bucketName: string,
  s3Key: string,
  html: aws.S3.Types.Body
) {
  const params: aws.S3.Types.PutObjectRequest = {
    Body: html,
    ACL: "public-read", // users can access email html previews
    Bucket: bucketName,
    CacheControl: "public, immutable",
    // https://stackoverflow.com/a/2505733/5093093
    ContentType: "text/html; charset=UTF-8",
    StorageClass: "REDUCED_REDUNDANCY",
    Key: s3Key,
  };

  await new aws.S3().putObject(params).promise();
}

File: shared/src/s3.rs
use async_trait::async_trait;
use futures::{StreamExt, TryStreamExt};
use rusoto_core::RusotoError;
use rusoto_s3::{
    GetObjectError, GetObjectRequest, PutObjectError, PutObjectRequest,
    S3Client, S3,
};
use serde::Deserialize;
use std::str::FromStr;

/// Implements only methods which this project requires instead of all
/// [`rusoto_s3::S3`] methods, which makes it more comfortable to write stubs
/// and test it.
#[async_trait]
pub trait S3Ext {
    async fn put(
        &self,
        bucket: String,
        key: String,
        body: Vec<u8>,
        conf: PutConf,
    ) -> Result<(), RusotoError<PutObjectError>>;

    async fn get(
        &self,
        bucket: String,
        key: String,
    ) -> Result<Option<Vec<u8>>, RusotoError<GetObjectError>>;
}

#[derive(Default, PartialEq, Debug)]
pub struct PutConf {
    pub acl: Option<String>,
    pub cache_control: Option<String>,
    pub content_type: Option<String>,
}

#[async_trait]
impl S3Ext for S3Client {
    async fn put(
        &self,
        bucket: String,
        key: String,
        body: Vec<u8>,
        conf: PutConf,
    ) -> Result<(), RusotoError<PutObjectError>> {
        let req = PutObjectRequest {
            acl: conf.acl,
            cache_control: conf.cache_control,
            content_type: conf.content_type,
            body: Some(body.into()),
            bucket,
            key,
            storage_class: Some("REDUCED_REDUNDANCY".to_string()),
            ..Default::default()
        };
        self.put_object(req).await?;
        Ok(())
    }

    async fn get(
        &self,
        bucket: String,
        key: String,
    ) -> Result<Option<Vec<u8>>, RusotoError<GetObjectError>> {
        let body = self
            .get_object(GetObjectRequest {
                bucket,
                key,
                ..Default::default()
            })
            .await?
            .body;

        if let Some(body) = body {
            let body: Vec<_> = body
                .into_stream()
                .collect::<Vec<_>>()
                .await
                .into_iter()
                .collect::<Result<Vec<_>, _>>()?
                .into_iter()
                .flatten()
                .collect();

            Ok(Some(body))
        } else {
            Ok(None)
        }
    }
}

/// Message published by S3 when a new object is created.
///
/// [https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NewS3Object {
    pub region: String,
    pub key: String,
    pub bucket: String,
}

impl FromStr for NewS3Object {
    type Err = serde_json::Error;

    fn from_str(message_body: &str) -> Result<Self, Self::Err> {
        let records =
            serde_json::from_str::<NewS3ObjectEvent>(message_body)?.records;

        assert_eq!(
            1,
            records.len(),
            "There must always be exactly one record in an SQS message"
        );
        let record = records.into_iter().next().unwrap();

        Ok(Self {
            region: record.aws_region,
            key: record.s3.object.key,
            bucket: record.s3.bucket.name,
        })
    }
}

#[derive(Deserialize)]
struct NewS3ObjectEvent {
    #[serde(rename = "Records")]
    records: Vec<NewS3ObjectEventRecord>,
}

#[derive(Deserialize)]
struct NewS3ObjectEventRecord {
    #[serde(rename = "awsRegion")]
    aws_region: String,
    s3: NewS3ObjectEventRecordS3,
}

#[derive(Deserialize)]
struct NewS3ObjectEventRecordS3 {
    bucket: NewS3ObjectEventRecordS3Bucket,
    object: NewS3ObjectEventRecordS3Object,
}

#[derive(Deserialize)]
struct NewS3ObjectEventRecordS3Bucket {
    name: String,
}

#[derive(Deserialize)]
struct NewS3ObjectEventRecordS3Object {
    key: String,
}
